channeldb/payments.go

```diff
@@ -66,265 +116,334 @@ func (ps *PaymentStatus) FromBytes(status []byte) error {
-func (db *DB) AddPayment(payment *OutgoingPayment) error {
-	// Validate the field of the inner voice within the outgoing payment,
-	// these must also adhere to the same constraints as regular invoices.
-	if err := validateInvoice(&payment.Invoice); err != nil {
-		return err
-	}
+// PaymentAttemptInfo contains information about a specific payment attempt for
+// a given payment. This information is used by the router to handle any errors
+// coming back after an attempt is made, and to query the switch about the
+// status of a payment. For settled payment this will be the information for
+// the succeeding payment attempt.
+type PaymentAttemptInfo struct {
+	// PaymentID is the unique ID used for this attempt.
+	PaymentID uint64
+
+	// Route is the route attempted to send the HTLC.
+	Route route.Route
+}
 
-	// We first serialize the payment before starting the database
-	// transaction so we can avoid creating a DB payment in the case of a
-	// serialization error.
-	var b bytes.Buffer
-	if err := serializeOutgoingPayment(&b, payment); err != nil {
-		return err
-	}
-	paymentBytes := b.Bytes()
+// Payment is a wrapper around a payment's PaymentCreationInfo,
+// PaymentAttemptInfo, and preimage. All payments will have the
+// PaymentCreationInfo set, the PaymentAttemptInfo will be set only if at least
+// one payment attempt has been made, while only completed payments will have a
+// non-zero payment preimage.
+type Payment struct {
+	// sequenceNum is a unique identifier used to sort the payments in
+	// order of creation.
+	sequenceNum uint64
 
-	return db.Batch(func(tx *bbolt.Tx) error {
-		payments, err := tx.CreateBucketIfNotExists(paymentBucket)
-		if err != nil {
-			return err
-		}
+	// Status is the current PaymentStatus of this payment.
+	Status PaymentStatus
 
-		// Obtain the new unique sequence number for this payment.
-		paymentID, err := payments.NextSequence()
-		if err != nil {
-			return err
-		}
+	// PaymentCreationInfo is populated when the payment is initiated.
+	*PaymentCreationInfo
 
-		// We use BigEndian for keys as it orders keys in
-		// ascending order. This allows bucket scans to order payments
-		// in the order in which they were created.
-		paymentIDBytes := make([]byte, 8)
-		binary.BigEndian.PutUint64(paymentIDBytes, paymentID)
+	// PaymentAttemptInfo is the information about the last payment attempt
+	// made.
+	//
+	// NOTE: Can be nil if no attempt is yet made.
+	*PaymentAttemptInfo
 
-		return payments.Put(paymentIDBytes, paymentBytes)
-	})
+	// PaymentPreimage is the preimage of a successful payment. This serves
+	// as a proof of payment. It will be all zeroes for non-settled
+	// payments.
+	PaymentPreimage lntypes.Preimage
 }
 
-// FetchAllPayments returns all outgoing payments in DB.
-func (db *DB) FetchAllPayments() ([]*OutgoingPayment, error) {
-	var payments []*OutgoingPayment
+// FetchPayments returns all sent payments found in the DB.
+func (db *DB) FetchPayments() ([]*Payment, error) {
```

Comment:
It seems this should be moved into `control_tower.go`.

Reply:
Opted for having this method directly on the db instead on the control tower since it is accessed by the `rpcserver`.

Reply:
In general I think we don't manipulate the same buckets from multiple structs. Is it a problem to give rpc server access to control tower?

Reply:
No big deal, it is just a wrapper around the DB anyway. I don't seen much to gain from doing it though. 

Maybe we should merge `payments.go` and `control_tower.go`?

Reply:
Yes I think that would make sense. Half of the code that control tower uses is in payments.

Reply:
Created follow-up: https://github.com/lightningnetwork/lnd/pull/3124